src/builder/pdb.rs

```diff
@@ -0,0 +1,311 @@
+use crate::{
+    builder::ObjectMetaBuilder,
+    error::OperatorResult,
+    labels::{role_selector_labels, APP_MANAGED_BY_LABEL},
+    utils::format_full_controller_name,
+};
+use k8s_openapi::{
+    api::policy::v1::{PodDisruptionBudget, PodDisruptionBudgetSpec},
+    apimachinery::pkg::{
+        apis::meta::v1::{LabelSelector, ObjectMeta},
+        util::intstr::IntOrString,
+    },
+};
+use kube::{Resource, ResourceExt};
+
+#[derive(Debug, Default)]
+/// This builder is used to construct [`PodDisruptionBudget`]s.
+/// If you are using this to create [`PodDisruptionBudget`]s according to ADR 30 on Pod disruptions,
+/// the use of [`PodDisruptionBudgetBuilder::new_with_role`] is recommended.
+///
+/// The following attributes on a [`PodDisruptionBudget`] are considered mandatory and must be specified
+/// before being able to construct the [`PodDisruptionBudget`]:
+///
+/// 1. `metadata`
+/// 2. `selector`
+/// 3. Either `minAvailable` or `maxUnavailable`
+///
+/// Both `metadata` and `selector` will be set by [`PodDisruptionBudgetBuilder::new_with_role`].
+pub struct PodDisruptionBudgetBuilder<ObjectMeta, LabelSelector, PodDisruptionBudgetConstraint> {
+    metadata: ObjectMeta,
+    selector: LabelSelector,
+    /// Tracks wether either `maxUnavailable` or `minAvailable` is set.
+    constraint: Option<PodDisruptionBudgetConstraint>,
+}
+
+/// We intentionally only support fixed numbers, no percentage, see ADR 30 on Pod disruptions for details.
+/// We use u16, as [`IntOrString`] takes an i32 and we don't want to allow negative numbers. u16 will always fit in i32.
+#[derive(Debug)]
+pub enum PodDisruptionBudgetConstraint {
+    MaxUnavailable(u16),
+    MinAvailable(u16),
+}
+
+impl PodDisruptionBudgetBuilder<(), (), ()> {
+    pub fn new() -> Self {
+        PodDisruptionBudgetBuilder::default()
+    }
+
+    /// This method populates `metadata` and `selector` from the give role (not roleGroup!).
+    ///
+    /// The parameters are the same as the fields from [`crate::labels::ObjectLabels`]:
+    /// * `owner` - Reference to the k8s object owning the PDB, this should be e.g. `HdfsCluster` or `TrinoCluster`.
```

Comment:
```suggestion
    /// * `owner` - Reference to the k8s object owning the PDB, for example `HdfsCluster` or `TrinoCluster`.
```

Reply:
changed in 9889d9bf02ee985998b3f8339cf9924da9891d9b